tools/c7n_mailer/c7n_mailer/slack_delivery.py

```diff
@@ -0,0 +1,112 @@
+
+            if not resolved_addrs:
+                continue
+
+            for address, slack_target in resolved_addrs.iteritems():
+                slack_messages[address] = get_rendered_jinja(slack_target, sqs_message, resources,
+                                                             self.logger, 'slack_template', 'slack_default')
+
+        return slack_messages
+
+    def slack_handler(self, sqs_message, slack_messages):
+
+        for message in slack_messages:
+            self.logger.info("Sending account:%s policy:%s %s:%s email:%s to %s" % (
+                                 sqs_message.get('account', ''),
+                                 sqs_message['policy']['name'],
+                                 sqs_message['policy']['resource'],
+                                 str(len(sqs_message['resources'])),
+                                 sqs_message['action'].get('slack_template', 'slack_default'),
+                                 json.loads(slack_messages[message])["channel"])
+                            )
+
+            self.send_slack_msg(slack_messages[message])
+
+    def retrieve_user_im(self, email_addresses):
+        list = {}
+        for address in email_addresses:
+            if self.cache_engine and self.caching.get(address):
+                    self.logger.debug('Got slack metadata from local cache for: %s' % address)
+                    list[address] = self.caching.get(address)
+                    continue
+
+            response = self.client.api_call(
+                "users.lookupByEmail", email=address)
+
+            if not response["ok"] and "Retry-After" in response["headers"]:
+                raise Exception("Slack API timeout.")
+            elif not response["ok"] and response["error"] == "invalid_auth":
+                raise Exception("Invalid Slack token.")
+                return
+            elif not response["ok"] and response["error"] == "users_not_found":
+                self.logger.info("Slack user ID not found.")
+                self.caching.set(address, {})
+                continue
+            else:
+                self.logger.debug("Slack account %s found for user %s", response['user']['enterprise_user']['id'])
+                if self.cache_engine:
+                    self.logger.debug('Writing user: %s metadata to cache engine.' % address)
+                    self.caching.set(address, response['user']['enterprise_user']['id'])
+
+                list[address] = response['user']['enterprise_user']['id']
+
+        return list
+
+    def send_slack_msg(self, message_payload):
+        response = requests.post(
+            url='https://slack.com/api/chat.postMessage',
+            data=message_payload,
+            headers={'Content-Type': 'application/json;charset=utf-8',
+                       'Authorization': 'Bearer %s' % self.config.get('slack_token')}
+        )
+
+        if response.status_code != 200:
```

Comment:
Should handle rate limiting here as well, https://api.slack.com/docs/rate-limits